Docs/Architecture_DAS.md

```diff
@@ -0,0 +1,141 @@
+
+#### Container Orchestration
+- **Kubernetes (K8s)**: DAS is containerized and managed using Kubernetes, which facilitates automated deployment, scaling, and management of containerized applications.
+- **Clusters and Nodes**: The Kubernetes cluster comprises multiple nodes that host the containerized microservices. Each service, such as the API Layer, Service Layer, and Data Layer, is deployed as a Kubernetes Pod.
+- **Horizontal Pod Autoscaling**: Kubernetes enables Horizontal Pod Autoscaling (HPA) based on CPU or memory utilization, dynamically scaling the number of pods to meet current demand.
+- **Service Discovery and Load Balancing**: Kubernetes services provide internal load balancing within the cluster and manage communication between pods, ensuring seamless and reliable inter-service communication.
+
+#### CI/CD Pipelines
+- **Pipeline Overview**: Continuous Integration and Continuous Deployment (CI/CD) pipelines are set up to automate building, testing, and deployment processes, reducing manual intervention and ensuring rapid, reliable delivery of updates.
+- **Tools**: The CI/CD pipelines are configured with tools such as Jenkins, GitLab CI/CD, or GitHub Actions for automated workflows, integrating with AWS and Kubernetes for seamless deployment.
+- **Build Stage**: The pipeline triggers builds upon code commits or pull requests, creating Docker images for each service in DAS and pushing them to a secure container registry (e.g., AWS Elastic Container Registry (ECR)).
+- **Test Stage**: Automated tests (unit, integration, and end-to-end) run in parallel to validate changes. Any failures stop the deployment process, ensuring only fully tested code reaches production.
+- **Deployment Stage**: Upon passing all tests, Kubernetes deployments are triggered, and updated containers are rolled out. Kubernetes Rolling Updates ensure zero-downtime deployment by gradually replacing instances without affecting live traffic.
+
+#### Monitoring and Logging
+- **Monitoring**: Amazon CloudWatch and Prometheus are used for monitoring system performance, tracking key metrics (CPU, memory, request rates), and identifying any issues early. Kubernetes also provides native metrics for pod health and resource usage.
+- **Logging**: AWS CloudWatch Logs or ELK Stack (Elasticsearch, Logstash, Kibana) captures and organizes logs for audit, security, and troubleshooting. Logs are centralized to provide insights across all microservices and help detect anomalies or performance bottlenecks.
+
+## 8. Cross-cutting Concepts
+### Security
+- **Authentication and Authorization**: Keycloak provides user authentication and authorization services.
+- **Data Encryption**: All sensitive data is encrypted in transit and at rest.
+- **Access Control**: Role-based access control (RBAC) is enforced for sensitive data.
+
```

Comment:
```suggestion
#### Security Architecture

Security in the DAS application is paramount and is built upon the following core principles:

*   **Decentralized and Declarative Access Control:** Access control is enforced directly within the service layer (e.g., `ledgers-account-service-impl`) and is entirely declarative. This means that access permissions are defined by associating accessor public keys with the resources they are allowed to access in a dedicated access control database. The presented credential serves as an attestation of this right, eliminating the need for real-time checks against the access database during authorization.

*   **Public Key-Based Authentication:**  DAS handles user authentication directly, relying on public key cryptography.  Accounts are linked to accessor public keys. Access is granted based on:
    *   **Signed Credentials (SD-JWT):** Issued by DAS to authenticated users, these credentials assert the user's identity and their associated public key.
    *   **Request Presentations:** For each request, the accessor must sign a presentation containing the request's hash. This presentation is included in the `Authorization` header.

*   **Request Signing and Verification:** DAS verifies the signature on the presentation to authenticate the requestor. This confirms that the request originates from the claimed accessor and hasn't been tampered with.

*   **Comprehensive Logging:** All requests, along with their associated presentations, are meticulously logged for auditing and security analysis. This creates a detailed audit trail for identifying potential security breaches or suspicious activities.

*   **Revocation Handling:**  A revocation database maintains a list of revoked credentials that haven't expired. This database is also held in memory for faster access, ensuring revoked credentials are promptly identified and rejected.

#### Data Encryption

Encryption remains crucial for protecting sensitive data:

*   **TLS/SSL for Transport Encryption:** All communication between clients and DAS, as well as inter-service communication, is secured using TLS/SSL to safeguard data in transit.
*   **Encryption at Rest:** Sensitive data in databases or persistent storage is encrypted to protect against unauthorized access in case of data breaches.

#### Diagram for Request Flow with Security

```mermaid
sequenceDiagram
    participant Client
    participant DAS

    Client->>DAS: Request with Signed Presentation
    activate DAS
    DAS->>DAS: Verify Presentation Signature
    DAS->>Service Layer: Access Control Check (Declarative)
    Service Layer-->>DAS: Access Granted
    DAS-->>Client: Response
    deactivate DAS
```

#### Diagram for Security Components

```mermaid
graph LR
    subgraph DAS
        A[Service Layer]
        B[Credential Issuance]
        C[Revocation DB]
        D[Logging]
    end
    A --> B
    A --> C
    A --> D
    Client --> A
```
```